### PR TITLE
Closes #3785 #3778 #3779 #3788

### LINKED ISSUE #3785: Implement Layer Filter Concept
This task captures the discussion we have had in the epic about implementing the possibility to implement the concept of **Layer Filter**.

A **Layer Filter** acts directly on a layer coming from GeoServer and filter its content upfront. Once a filter is created for a layer:

- the map content for this layer is automatically filtered
- if we issue a GetFeatureInfo, the features we get are automatically filtered using this filter
- if we open the FeatureGrid for this layer, the features we get are automatically filtered using this filter. Any additional filter we define using the Query Builder or the Quick Filters will go in AND with the layer filter. 
- if we create a Widget from this layer, the features we get are automatically filtered using this filter. Any additional filter we define using the Query Builder or the Quick Filters will go in AND with the layer filter.
- the filter will be persisted along with the other layer options, inside the map object on map save. If we reload the map once it's saved the filter will be reapplied.

A filter icon shall be shown  in the TOC right next to the layer title to notify that the layer is filtered. The Filter Icon should be toggable to temporarily disable the filter from anywhere above.

### PR DESCRIPTION
**Pull request by kappu72**
## Description
With this pull a new Layer filter concept is added.
The user is able to set up a filter at layer level that will be used in any context (FG, QB, WMS Plugin)
see #3785 issue.
It also closes some minor bugs

## Issues
 -  #3785 
 - #3778
- #3779
- #3788
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)

 - [x] Bugfix
 - [ x] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)
It's not possible to define a filter at layer level

**What is the new behavior?**
It become possible to define a filter at layer level

**Does this PR introduce a breaking change?** (check one with "x", remove the other)

 - [ ] Yes, and I documented them in migration notes
 - [X] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: web/client/actions/filterHistory.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+ * LICENSE file in the root directory of this source tree.
+*/
+
+const INIT_FILTER_HISTORY = 'FILTER_HISTORY:INIT_FILTER_HISTORY';
+const APPLIED_FILTER = 'FILTER_HISTORY:APPLIED_FILTER';
+const STORE_CURRENT_APPLIED_FILTER = 'FILTER_HISTORY:STORE_CURRENT_APPLIED_FILTER';
+const RESTORE_CURRENT_SAVED_FILTER = 'FILTER_HISTORY:RESTORE_CURRENT_SAVED_FILTER';
+const APPLY_FILTER = 'FILTER_HISTORY:APPLY_FILTER';
+const OPEN_QUERY_BUILDER = 'LAYER_FILTER:OPEN_QUERY_BUILDER';
+
```
Comment (offtherailz) [inline web/client/actions/filterHistory.js]: You should document a little bit the actions to explain their role. So you can find better names for actions

File: web/client/epics/layerfilter.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    return changeLayerProperties(layer, {layerFilter: filter});
+};
+// Create action to add nativeCrs to wms layer
+const addNativeCrsLayer = (layer, nativeCrs = "") => {
+    return changeLayerProperties(layer, {nativeCrs});
+};
+
+
+module.exports = {
+    handleLayerFilterPanel: (action$, {getState}) =>
```
Comment (offtherailz) [inline web/client/epics/layerfilter.js]: Document the epics to explain what they do.

File: web/client/actions/filterHistory.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+*/
+
+const INIT_FILTER_HISTORY = 'FILTER_HISTORY:INIT_FILTER_HISTORY';
+const APPLIED_FILTER = 'FILTER_HISTORY:APPLIED_FILTER';
+const STORE_CURRENT_APPLIED_FILTER = 'FILTER_HISTORY:STORE_CURRENT_APPLIED_FILTER';
+const RESTORE_CURRENT_SAVED_FILTER = 'FILTER_HISTORY:RESTORE_CURRENT_SAVED_FILTER';
+const APPLY_FILTER = 'FILTER_HISTORY:APPLY_FILTER';
+const OPEN_QUERY_BUILDER = 'LAYER_FILTER:OPEN_QUERY_BUILDER';
```
Comment (offtherailz) [inline web/client/actions/filterHistory.js]: These actions could be part of the queryform reducer and actions. FilterHistory is anyway a bad name. See my comment about history, anyway

Comment (kappu72) [reply web/client/actions/filterHistory.js]: I prefer to keep separate, queryform reducer/actions are already enough long files full of actions. 
As suggested I have modified the name.

**Comment by kappu72** [comment]
Minutes of call with Lorenzo about his comments.
We needs to clarify if discard btn should be active only after a filter is changed (apply btn pressed) or after any changes to the filter builder ui.
The reset btn has just to reset filter and not to store it.
Save btn has to store the filter but not to close the panel.

**Comment by offtherailz** [comment]
So as @tdipisa asked, I'm going to merge. @kappu72 if you already worked on fixes, please do a separate pull request to add them.

**Pull request merged**